### PR TITLE
add worker debugger port to the state support

### PR DESCRIPTION
**Pull request by sofianhnaide**
## Why are these changes needed?
Add `debugger_port` to the state API, this will allow ray to track which worker have a debugger server running and what port is open on those worker.

## Manual test

I ran the following code.
```
import ray
ray.init()

@ray.remote
def test_task():
    print("Worker id: ", ray.WorkerID(ray._private.worker.global_worker.worker_id))
    ray._private.worker.global_worker.set_debugger_port(12345)
    import time
    time.sleep(60)
    return 1


result = test_task.remote()
print(ray.get(result))
```

and then did: 
```
(base) ray@g-e1d12ab7226970001:~/default$ ray get workers bf9e0104210d9ae55220c0740453adbd25d0e54e563de9ba09bf3445
---
-   worker_id: bf9e0104210d9ae55220c0740453adbd25d0e54e563de9ba09bf3445
    is_alive: true
    worker_type: WORKER
    exit_type: null
    node_id: 44f3b738208b5f78cccb50404046cef0ee632a2a0514205e60d22926
    ip: 10.0.0.102
    pid: 213374
    exit_detail: null
    worker_launch_time_ms: '2023-11-06 12:32:10.526000'
    worker_launched_time_ms: '2023-11-06 12:32:11.380000'
    start_time_ms: '2023-11-06 12:32:11.414000'
    end_time_ms: '1969-12-31 16:00:00'
    debugger_port: 12345
...
```

and verified that the debugger_port is assigned.
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/state.py [hunk 1]
```diff
@@ -664,6 +664,21 @@ def add_worker(self, worker_id, worker_type, worker_info):
             worker_data.SerializeToString()
         )
 
+    def update_worker_debugger_port(self, worker_id, debugger_port):
```
Comment (pcmoritz) [inline python/ray/_private/state.py:667]: Consider adding an underscore here and for `update_worker_debugger_port` below so it doesn't accidentially end up in a public API / people's autocomplete :)

**Review by pcmoritz — APPROVED** [review]
It looks great to me :)

File: python/ray/_private/state.py (hunk 1 above)
Comment (sofianhnaide) [reply python/ray/_private/state.py:667]: actually looking into this more closely, is this part of the public API. When I look state I find those API's which is differently: https://docs.ray.io/en/latest/ray-observability/reference/api.html

I am working under the expectation that this should only be used in ray internals.

Comment (rickyyx) [reply python/ray/_private/state.py:667]: I think we expose a few from this `state.py` as public APIS (developers): https://github.com/ray-project/ray/blob/c051a01953230c48cf8ad77edc288b75e7b9ab45/python/ray/__init__.py#L105-L110

File: python/ray/_private/worker.py [hunk 2]
```diff
@@ -541,6 +543,16 @@ def runtime_env(self):
         """Get the runtime env in json format"""
         return self.core_worker.get_current_runtime_env()
 
+    @property
+    def debugger_port(self):
+        """Get the debugger port for this worker"""
+        return self._debugger_port
+
+    def set_debugger_port(self, port):
+        worker_id = self.core_worker.get_worker_id()
```
Comment (rickyyx) [inline python/ray/_private/worker.py:552]: nit: I wonder if we just want to embed the `update_worker_debugger_port` implementation here in the `worker.py` file. 

E.g. through `ray._private.state.state.global_state_accessror` like:

https://github.com/ray-project/ray/blob/70dbcbffddf0bf05f48d27f89f981951672cee34/python/ray/_private/worker.py#L2627-L2635

File: python/ray/_private/state.py [hunk 3]
```diff
@@ -933,3 +948,16 @@ def available_resources():
             resource in the cluster.
     """
     return state.available_resources()
+
+
+def update_worker_debugger_port(worker_id, debugger_port):
```
Comment (rickyyx) [inline python/ray/_private/state.py:953]: assert `debugger_port` not None?

**Review by rickyyx — APPROVED** [review]
Looks great!!

File: python/ray/_private/worker.py (hunk 2 above)
Comment (sofianhnaide) [reply python/ray/_private/worker.py:552]: I will defer this to you and others in the core team to give the direction, I don't think we want this to be an external API and I do see few usage `ray._private.state`, so not sure what the difference here between the two.

File: python/ray/_private/state.py (hunk 3 above)
Comment (sofianhnaide) [reply python/ray/_private/state.py:959]: done

File: python/ray/_private/state.py (hunk 1 above)
Comment (sofianhnaide) [reply python/ray/_private/state.py:667]: great, so I will keep as is here since it's not exposed by default

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
retrying the window build...

**Comment by rkooo567** [comment]
@sofianhnaide can you ping me when the new window wheel build finishes and doesn't fail?

**Comment by sofianhnaide** [comment]
> @sofianhnaide can you ping me when the new window wheel build finishes and doesn't fail?

rebasing

**Pull request merged**